### ISSUE TITLE
Test Suite: xfstests build issue

### ISSUE THREAD

**Issue by behlendorf** [zfsonlinux/zfs#5481]
Recent updates to CentOS 7, Ubuntu 16.04, and almost certain other modern versions of Linux have resulted in [xfstests no longer building](http://build.zfsonlinux.org/builders/CentOS%207%20x86_64%20Mainline%20%28TEST%29/builds/740/steps/shell_10/logs/make).  The failures have been resolved in current versions of xfstests.  What needs to happen is for our patch stack to be rebased on the latest xfstests and then submitted upstream for inclusion.

**Comment by implr** [comment]
I might take a stab at rebasing this over a weekend. It's a bit hard to piece together what is the current state though. Can you confirm that branch `zfs` of https://github.com/zfsonlinux/xfstests is the latest version of the patchset, or does someone somewhere have something further along?

**Comment by behlendorf** [comment]
@implr that would be very welcome.  That is the latest version of the branch I'm aware of, I believe xfstests has changed considerably side it was last updated so you might want to use those commits as a guide of what needs to be updated to get things going.  I'd love to be able to add xfstests to our regular testing.

**Comment by implr** [comment]
Yeah I had a brief look already, rebasing those patches over a decade of changes is hopeless. From scratch shouldn't be too bad though.
Another thing is that xfstests gained a bunch of zfs-relevant tests because of btrfs (reflinks, dedup, etc). Those might be a bit harder to get working due to btrfs specifics, but making generic run looks very doable.

**Comment by implr** [comment]
Small progress update: I have a partial patch at https://github.com/implr/xfstests/tree/zfs. I tested on ubuntu2204 for now with their packaged zfs, so I didn't try any  of the reflink/FICLONE tests yet. I used this local.config:
```
export TEST_DEV=test/t
export TEST_DIR=/mnt/test
export SCRATCH_DEV=/dev/vdc
# will be destroyed/recreated, pick a name that doesn't exist on your system
export SCRATCH_ZPOOL_NAME=scrat
export SCRATCH_MNT=/mnt/scratch
```
Note SCRATCH_DEV is a block device. I kept @behlendorf's "TEST_DEV is a fs name" approach for convenience, but I really wanted scratch to be a blkdev, as that would allow us to try the various dm-flakey tests (and perhaps even the btrfs pool tests). However, that ended up being quite complicated, as ZFS pools are identified by names, while a lot of parts of xfstests really want to be able to just `mount $SCRATCH_DEV`. I fixed it for the general case, but a lot of the devmapper stuff remains broken.

A bunch of other tests fail too. Some of them look like I'm missing things in the port, some are more suspicious, but I didn't have time to debug them yet. `exclude.zfs.txt` has the list.

**Comment by implr** [comment]
I've moved onto getting the reflink tests to run and ran into a problem.
xfstests checks if the fs supports clone_range [like this](https://github.com/implr/xfstests/blob/zfs/common/reflink#L52). That always fails on zfs for two reasons:
1. 64k < default recordsize, so it fails with EINVAL
2. it creates and then clones the file immediately after, which usually falls within the same txg, so zfs returns EAGAIN (and xfs_io doesn't retry)

While 1 is trivial to work around (lower recordsize on the test fs, or just patch the 64k to something bigger), I don't really know what to do about 2. For a proper fix we could patch xfs_io, but that might take a while to upstream, and meanwhile be tedious to manage (it's a separate thing, not part of xfstests). A workaround would be waiting for the next txg, but how? I guess there's always an option for a few second sleep, but that's really stupid.

Any ideas?

**Comment by behlendorf** [comment]
#15842 should be the fix for 2.  It would be particularly helpful if you could run the reflink tests with that PR.  Hopefully it'll help us identify any other places where ZFS behaves differently than the other Linux filesystem which support reflink.

**Comment by implr** [comment]
Hello again, it's been a while.
I got burned out by this 2 years ago, the extremely long iteration times plus the high likelihood of making dumb errors (because everything is badly documented bash) made cleaning it up quite unpleasant. Fortunately nowadays we can have LLMs slog through that. 

I rebased on top of current (as of ~March) upstream xfstests and had Claude chase down the remaining issues. The whole auto group now runs, with exceptions in `exclude.zfs.txt`. It's a mix - some are almost definitely bugs in zfs (mostly lockups), some are (often expected) behavioral differences, and the rest could be made to run with (lots of) more effort (eg. they construct a layered dm/loop contraption which would need to be plumbed to the vdevs).
The current state is [implr/xfstests/zfs](https://github.com/implr/xfstests/tree/zfs) as before.

A note on code quality:
I reviewed the implementation itself and I stand by it. I wrote a good chunk of it myself in the first attempt, there's a couple ugly hacks but otherwise I think it's decent.
The root causes in exclude comments come from Claude. The tests do actually fail, but the reasoning why might be wrong. Needs proper investigation.
Commit messages (and some code comments) are kinda slop. I'll probably just squash them, as I don't think there's much value there.

@behlendorf Where do we go from here?
Would be nice to have this run as CI. I can put something together, but probably can't run it long term.
Another thing is chasing down the actual bugs. I might take a stab at something simple soon.

**Comment by behlendorf** [comment]
@implr welcome back!  Thanks for picking this up again and pushing it forward.

As for next steps, I think getting what you have up and running as a GitHub Actions based workflow is the way to go.  I'd suggest making a copy of the `zfs-qemu.yml` workflow to start with and then updating it to run xfstests for a single distribution. That way you'll be able to leverage the existing support scripts to bootstrap a qemu test environment with console logging.  Once you have that we should be able to pull some version of it in to the OpenZFS repository which we can trigger manually, and then start iterating on that.

> Commit messages (and some code comments) are kinda slop. I'll probably just squash them

That's fine.  We should probably also think about forking a version of the git repository in to the openzfs organization.  But there's no rush there.
